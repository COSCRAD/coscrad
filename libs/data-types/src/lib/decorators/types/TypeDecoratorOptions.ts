export type TypeDecoratorOptions = {
    isOptional: boolean;

    isArray: boolean;
};
