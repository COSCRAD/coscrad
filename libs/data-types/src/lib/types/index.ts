export * from './ClassDataTypeMetadata';
export * from './CoscradDataSchema';
export * from './CoscradDataType';
export * from './PropertyTypeMetadata';
