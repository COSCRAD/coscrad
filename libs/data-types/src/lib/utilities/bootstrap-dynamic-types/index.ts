export * from './bootstrap-dynamic-types';
