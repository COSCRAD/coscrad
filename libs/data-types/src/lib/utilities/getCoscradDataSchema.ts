import { ClassSchema } from '../types/ClassSchema';
import getCoscradDataSchemaFromPrototype from './getCoscradDataSchemaFromPrototype';

// eslint-disable-next-line
export default (TargetClass: Object): ClassSchema<Record<string, unknown>> =>
    // @ts-expect-error TODO: restrict argument to be a class
    getCoscradDataSchemaFromPrototype(TargetClass.prototype || {});
