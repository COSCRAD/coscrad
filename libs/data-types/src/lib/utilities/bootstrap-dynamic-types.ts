import { isNonEmptyString, isNullOrUndefined } from '@coscrad/validation-constraints';
import { COSCRAD_DATA_TYPE_METADATA } from '../constants';
import {
    getUnionMemberMetadata,
    getUnionMetadata,
    isUnionMemberMetadata,
    isUnionMetadata,
} from '../decorators';
import getCoscradDataSchema from './getCoscradDataSchema';
import { Ctor } from './getCoscradDataSchemaFromPrototype';

type ComplexCoscradDataTypeDefinition = {
    complexDataType: string;
};

// We duplicate this here to avoid circular build dependencies
const isComplexCoscradDataTypeDefinition = (
    input: unknown
): input is ComplexCoscradDataTypeDefinition =>
    isNonEmptyString((input as ComplexCoscradDataTypeDefinition).complexDataType);

export const bootstrapDynamicTypes = (allCtorCandidates: unknown[]) => {
    const unionMetadata = allCtorCandidates.map(getUnionMetadata).filter(isUnionMetadata);
    // .filter(({ unionName }) => unionName === nameOfUnionToValidate);

    const unionNames = [...new Set(unionMetadata.map(({ unionName }) => unionName))];

    const unionMap = new Map<string, Map<string | number, Ctor<Object>>>();

    unionNames.forEach((nameOfUnionToValidate) => {
        const allDiscriminantPathsRegisteredForThisUnion = [
            ...new Set(unionMetadata.map(({ discriminantPath }) => discriminantPath)),
        ];

        // TODO It would be nice to design this away
        if (allDiscriminantPathsRegisteredForThisUnion.length > 1) {
            throw new Error(
                `You have defined more than one union data-type with the name: ${nameOfUnionToValidate}`
            );
        }

        if (allDiscriminantPathsRegisteredForThisUnion.length < 1) {
            throw new Error(
                `Failed to find a union data-type with the name: ${nameOfUnionToValidate}`
            );
        }

        const { discriminantPath } = unionMetadata[0];

        if (discriminantPath.includes('.')) {
            const msg = [
                `Using a nested property`,
                `as a discriminant is not supported.`,
                `Decorate the property on the nested data class instead`,
            ].join(' ');

            throw new Error(msg);
        }

        const discriminantValuesAndCtors = allCtorCandidates
            .map(getUnionMemberMetadata)
            .filter(isUnionMemberMetadata)
            .filter(({ unionName }) => nameOfUnionToValidate === unionName)
            .map(({ ctor, discriminantValue }) => [ctor, discriminantValue]);

        if (discriminantValuesAndCtors.length === 0) {
            // TODO Break out proper exceptions here and test for these
            throw new Error(
                `Failed to find any union members (sub-type classes) for the union: ${nameOfUnionToValidate}`
            );
        }

        const duplicateDiscriminants = discriminantValuesAndCtors
            .map(([_ctor, discriminantValue]) => discriminantValue)
            .reduce(
                (acc: { unique: unknown[]; duplicates: unknown[] }, discriminantValue) => {
                    const { unique, duplicates } = acc;

                    if (duplicates.includes(discriminantValue)) return acc;

                    if (unique.includes(discriminantValue))
                        return {
                            unique,
                            duplicates: [...duplicates, discriminantValue],
                        };

                    return {
                        duplicates,
                        unique: [...unique, discriminantValue],
                    };
                },
                {
                    unique: [],
                    duplicates: [],
                }
            ).duplicates;

        if (duplicateDiscriminants.length > 0) {
            throw new Error(
                `The following values are reused as union data-type discriminants: ${[
                    ...new Set(duplicateDiscriminants),
                ].join(', ')}`
            );
        }

        const membersMapForThisUnion = discriminantValuesAndCtors.reduce(
            (acc, [ctor, discriminantValue]: [Ctor<Object>, string | number]) =>
                acc.set(discriminantValue, ctor),
            new Map<string | number, Ctor<Object>>()
        );

        unionMap.set(nameOfUnionToValidate, membersMapForThisUnion);
    });

    /**
     * Now we iterate through all classes with `CoscradDataType` definitions,
     * writing the schemas for each union property.
     */
    const ctorsWithTypeDefinitions = allCtorCandidates.filter(
        (input) => !isNullOrUndefined(getCoscradDataSchema(input))
    );

    ctorsWithTypeDefinitions.forEach((c) => {
        const dataSchema = getCoscradDataSchema(c);

        const updates = Object.entries(dataSchema).reduce(
            (acc, [propertyKey, dataSchemaForProperty]) => {
                if (
                    isComplexCoscradDataTypeDefinition(dataSchemaForProperty) &&
                    dataSchemaForProperty.complexDataType === 'UNION_TYPE'
                ) {
                    const { unionName } = dataSchemaForProperty;

                    if (!unionMap.has(unionName)) {
                        /**
                         * This situation is impossible by design. The only way
                         * for the union name to appear on a data schema is to
                         * register the union. We add this logic out of an
                         * abundance of caution.
                         */
                        throw new Error(
                            `The union: ${unionName} that appears on property: ${propertyKey} of ${dataSchemaForProperty.label} has not been registered `
                        );
                    }

                    const unionMemberSchemaDefinitons = unionMap.get(unionName);

                    const unionMemberSchemaDefinitions = [
                        ...unionMemberSchemaDefinitons.entries(),
                    ].map(([discriminantValue, ctor]) => ({
                        discriminant: discriminantValue,
                        schema: getCoscradDataSchema(ctor),
                    }));

                    return {
                        ...acc,
                        [propertyKey]: {
                            schemaDefinitions: unionMemberSchemaDefinitions,
                        },
                    };
                } else {
                    /**
                     * no-op
                     *
                     * Note that it is not necessary to recurse as every class
                     * using data decorators should be exposed to the IoC
                     * registry.
                     */
                    return acc;
                }
            },
            {} as Record<string, { schemaDefinitions: unknown }>
        );

        Object.entries(updates).forEach(
            ([propertyKey, { schemaDefinitions: unionDataSchemasForThisProperty }]) => {
                const existingMeta = getCoscradDataSchema(c);

                const newMeta = {
                    ...existingMeta,
                    [propertyKey]: {
                        ...existingMeta[propertyKey],
                        schemaDefinitions: unionDataSchemasForThisProperty,
                    },
                };

                Reflect.defineMetadata(
                    COSCRAD_DATA_TYPE_METADATA,
                    newMeta,
                    // @ts-expect-error TODO fix me
                    c.prototype
                );
            }
        );
    });

    /**
     * Now we iterate through all known classes that leverage a union,
     * setting the schema from the union members.
     */

    // const allClassesLeveragingAUnion = allCtorCandidates.filter((input) =>
    //     isUnionMemberMetadata(getUnionMemberMetadata(input))
    // );

    // allClassesLeveragingAUnion.forEach((classReference) => {
    //     const { unionName } = getUnionMetadata(classReference) as UnionMetadata;

    //     // We already have checked for this this above

    //     const existingTypeMeta = Reflect.getMetadata(
    //         COSCRAD_DATA_TYPE_METADATA,
    //         // @ts-expect-error TODO fix types
    //         classReference.prototype
    //     );

    //     if (!existingTypeMeta) {
    //         throw new Error(
    //             `There is no COSCRAD data-type metadata for class: ${classReference} that leverages a union: ${unionName}`
    //         );
    //     }

    //     const schemaDefinitions = unionMap.get(unionName).forEach((ctor, discriminantValue) => ({
    //         discriminant: discriminantValue,
    //         schema: getCoscradDataSchema(ctor),
    //     }));

    //     Reflect.defineMetadata(
    //         COSCRAD_DATA_TYPE_METADATA,
    //         {
    //             ...existingTypeMeta,
    //             schemaDefinitions,
    //         },
    //         classReference
    //     );
    // });
};
