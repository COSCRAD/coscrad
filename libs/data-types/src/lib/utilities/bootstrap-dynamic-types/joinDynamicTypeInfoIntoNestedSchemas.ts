import { COSCRAD_DATA_TYPE_METADATA } from '../../constants';
import {
    ClassSchema,
    ComplexCoscradDataType,
    isSimpleCoscradPropertyTypeDefinition,
    NestedTypeDefinition,
} from '../../types';
import getCoscradDataSchema from '../getCoscradDataSchema';

type CoscradDataSchema = ReturnType<typeof getCoscradDataSchema>;

const joinDynamicTypeInfoIntoSchemaForANestedDataType = (
    schemaForNestedProperty: NestedTypeDefinition
): NestedTypeDefinition => {
    const { complexDataType, schema: originalSchema } = schemaForNestedProperty;

    if (complexDataType !== 'NESTED_TYPE') {
        throw new Error(`Expected a nested data type definition, received: ${complexDataType}`);
    }

    const updatedSchema = Object.entries(originalSchema).reduce(
        (acc, [propertyKey, propertyTypeDefinition]) => {
            if (isSimpleCoscradPropertyTypeDefinition(propertyTypeDefinition))
                return {
                    ...acc,
                    [propertyKey]: propertyTypeDefinition,
                };

            const { complexDataType } = propertyTypeDefinition;

            if (complexDataType === 'ENUM') {
                return {
                    ...acc,
                    [propertyKey]: propertyTypeDefinition,
                };
            }

            if (complexDataType === 'UNION_TYPE') {
                throw new Error(`Cannot handle a union type definition on a nested data type`);
            }

            if (complexDataType === ComplexCoscradDataType.nested) {
                return {
                    ...acc,
                    [propertyKey]:
                        // @ts-expect-error fix types
                        joinDynamicTypeInfoIntoSchemaForANestedDataType(propertyTypeDefinition),
                };
            }
        },
        {} as unknown as ClassSchema
    );

    return {
        complexDataType,
        // @ts-expect-error fix types
        schema: updatedSchema,
    };
};

// Doesn't this lib use camel case for naming files?
export const joinDynamicTypeInfoIntoNestedSchemas = (c: unknown): void => {
    const originalDataSchema = getCoscradDataSchema(c);

    const updatedDataSchema = Object.entries(originalDataSchema).reduce(
        (acc, [propertyKey, propertySchema]) => {
            if (
                isSimpleCoscradPropertyTypeDefinition(propertySchema) ||
                propertySchema.complexDataType !== ComplexCoscradDataType.nested
            )
                return {
                    ...acc,
                    [propertyKey]: propertySchema,
                };

            // We know we have a nested data type definition here
            return {
                ...acc,
                [propertyKey]: {
                    ...propertySchema,
                    // @ts-expect-error todo fix types (deal with mismatch between interface and concrete types here)
                    schema: joinDynamicTypeInfoIntoSchemaForANestedDataType(propertySchema),
                },
            };
        },
        {}
    );

    // @ts-expect-error fix errors here
    Reflect.defineMetadata(COSCRAD_DATA_TYPE_METADATA, updatedDataSchema, c.prototype);
};
