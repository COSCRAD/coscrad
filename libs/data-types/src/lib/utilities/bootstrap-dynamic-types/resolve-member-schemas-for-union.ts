import { UnionMemberMetadata } from '../../decorators';
import { ClassSchema, isSimpleCoscradPropertyTypeDefinition } from '../../types';
import getCoscradDataSchema from '../getCoscradDataSchema';
import { buildUnionTypesMap } from './bootstrap-dynamic-types';
import { leveragesUniontype } from './leverages-union-type';

export type UnionMemberSchemaDefinition = Pick<UnionMemberMetadata, 'discriminantValue'> & {
    schema: ClassSchema;
};

export const resolveMemberSchemasForUnion = (
    allCtorCandidates: unknown[],
    unionName: string
): UnionMemberSchemaDefinition[] => {
    const unionMap = buildUnionTypesMap(allCtorCandidates);

    const membersMap = unionMap.get(unionName);

    if (!membersMap) {
        return [];
    }

    return Array.from(membersMap.keys()).map((discriminantValue) => {
        const ctor = membersMap.get(discriminantValue);

        const rawSchema = getCoscradDataSchema(ctor);

        const updatedSchema = Object.entries(rawSchema).reduce(
            (acc, [propertyKey, propertyTypeDefinition]) => {
                if (
                    isSimpleCoscradPropertyTypeDefinition(propertyTypeDefinition) ||
                    !['NESTED_TYPE', 'UNION_TYPE'].includes(propertyTypeDefinition.complexDataType)
                )
                    return {
                        ...acc,
                        [propertyKey]: propertyTypeDefinition,
                    };

                const { complexDataType } = propertyTypeDefinition;

                if (complexDataType === 'NESTED_TYPE') {
                    const { schema: nestedSchema } = propertyTypeDefinition;

                    // @ts-expect-error fix types
                    if (!leveragesUniontype(nestedSchema))
                        return {
                            ...acc,
                            [propertyKey]: propertyTypeDefinition,
                        };

                    throw new Error(
                        `not implemented: cannot resolve union type: ${unionName} as one of its members leverages another union`
                    );
                }

                if (complexDataType === 'UNION_TYPE')
                    return {
                        ...acc,
                        [propertyKey]: {
                            ...propertyTypeDefinition,
                            schemaDefinitions: resolveMemberSchemasForUnion(
                                allCtorCandidates,
                                propertyTypeDefinition.unionName
                            ),
                        },
                    };
            },
            {}
        );

        return {
            discriminantValue,
            schema: updatedSchema,
        };
    });
};
