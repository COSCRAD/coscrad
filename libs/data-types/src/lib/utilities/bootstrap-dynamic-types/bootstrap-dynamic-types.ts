import { isNonEmptyString, isNullOrUndefined } from '@coscrad/validation-constraints';
import { COSCRAD_DATA_TYPE_METADATA } from '../../constants';
import {
    getUnionMemberMetadata,
    getUnionMetadata,
    isUnionMemberMetadata,
    isUnionMetadata,
} from '../../decorators';
import { isSimpleCoscradPropertyTypeDefinition } from '../../types';
import getCoscradDataSchema from './../getCoscradDataSchema';
import { Ctor } from './../getCoscradDataSchemaFromPrototype';
import { leveragesUniontype } from './leverages-union-type';
import { resolveMemberSchemasForUnion } from './resolve-member-schemas-for-union';

type ComplexCoscradDataTypeDefinition = {
    complexDataType: string;
};

export type UnionTypesMap = Map<string, Map<string | number, Ctor<Object>>>;

// We duplicate this here to avoid circular build dependencies
const isComplexCoscradDataTypeDefinition = (
    input: unknown
): input is ComplexCoscradDataTypeDefinition =>
    isNonEmptyString((input as ComplexCoscradDataTypeDefinition).complexDataType);

export const buildUnionTypesMap = (allCtorCandidates: unknown[]): UnionTypesMap => {
    const unionMetadata = allCtorCandidates.map(getUnionMetadata).filter(isUnionMetadata);

    const unionNames = [...new Set(unionMetadata.map(({ unionName }) => unionName))];

    const unionMap: UnionTypesMap = new Map();

    unionNames.forEach((nameOfUnionToValidate) => {
        const allDiscriminantPathsRegisteredForThisUnion = [
            ...new Set(unionMetadata.map(({ discriminantPath }) => discriminantPath)),
        ];

        // TODO It would be nice to design this away
        if (allDiscriminantPathsRegisteredForThisUnion.length > 1) {
            throw new Error(
                `You have defined more than one union data-type with the name: ${nameOfUnionToValidate}`
            );
        }

        if (allDiscriminantPathsRegisteredForThisUnion.length < 1) {
            throw new Error(
                `Failed to find a union data-type with the name: ${nameOfUnionToValidate}`
            );
        }

        const { discriminantPath } = unionMetadata[0];

        if (discriminantPath.includes('.')) {
            const msg = [
                `Using a nested property`,
                `as a discriminant is not supported.`,
                `Decorate the property on the nested data class instead`,
            ].join(' ');

            throw new Error(msg);
        }

        const discriminantValuesAndCtors = allCtorCandidates
            .map(getUnionMemberMetadata)
            .filter(isUnionMemberMetadata)
            .filter(({ unionName }) => nameOfUnionToValidate === unionName)
            .map(({ ctor, discriminantValue }) => [ctor, discriminantValue]);

        if (discriminantValuesAndCtors.length === 0) {
            // TODO Break out proper exceptions here
            throw new Error(
                `Failed to find any union members (sub-type classes) for the union: ${nameOfUnionToValidate}`
            );
        }

        const duplicateDiscriminants = discriminantValuesAndCtors
            .map(([_ctor, discriminantValue]) => discriminantValue)
            .reduce(
                (acc: { unique: unknown[]; duplicates: unknown[] }, discriminantValue) => {
                    const { unique, duplicates } = acc;

                    if (duplicates.includes(discriminantValue)) return acc;

                    if (unique.includes(discriminantValue))
                        return {
                            unique,
                            duplicates: [...duplicates, discriminantValue],
                        };

                    return {
                        duplicates,
                        unique: [...unique, discriminantValue],
                    };
                },
                {
                    unique: [],
                    duplicates: [],
                }
            ).duplicates;

        if (duplicateDiscriminants.length > 0) {
            throw new Error(
                `The following values are reused as union data-type discriminants: ${[
                    ...new Set(duplicateDiscriminants),
                ].join(', ')}`
            );
        }

        const membersMapForThisUnion = discriminantValuesAndCtors.reduce(
            (acc, [ctor, discriminantValue]: [Ctor<Object>, string | number]) =>
                acc.set(discriminantValue, ctor),
            new Map<string | number, Ctor<Object>>()
        );

        unionMap.set(nameOfUnionToValidate, membersMapForThisUnion);
    });

    return unionMap;
};

export const bootstrapDynamicTypes = (allCtorCandidates: unknown[]) => {
    /**
     * Now we iterate through all classes with `CoscradDataType` definitions,
     * writing the schemas for each union property.
     */
    const ctorsWithTypeDefinitions = allCtorCandidates.filter(
        (input) => !isNullOrUndefined(getCoscradDataSchema(input))
    );

    const ctorsThatLeverageAUnionType = ctorsWithTypeDefinitions.filter((ctor) =>
        leveragesUniontype(getCoscradDataSchema(ctor))
    );

    ctorsThatLeverageAUnionType.forEach((ctor) => {
        const dataSchema = getCoscradDataSchema(ctor);

        const updatedSchema = Object.entries(dataSchema).reduce(
            (acc, [propertyKey, propertyTypeDefinition]) => {
                if (
                    isSimpleCoscradPropertyTypeDefinition(propertyTypeDefinition) ||
                    !['NESTED_TYPE', 'UNION_TYPE'].includes(propertyTypeDefinition.complexDataType)
                )
                    return {
                        ...acc,
                        [propertyKey]: propertyTypeDefinition,
                    };

                const { complexDataType } = propertyTypeDefinition;

                if (complexDataType === 'NESTED_TYPE') {
                    throw new Error(`not implemented: we cannot yet bootstrap nested data types`);
                }

                if (complexDataType === 'UNION_TYPE') {
                    const { unionName } = propertyTypeDefinition;

                    return {
                        ...acc,
                        schemaDefinitions: resolveMemberSchemasForUnion(
                            allCtorCandidates,
                            unionName
                        ),
                    };
                }
            },
            {}
        );

        Reflect.defineMetadata(COSCRAD_DATA_TYPE_METADATA, updatedSchema, ctor.constructor);
    });
};
