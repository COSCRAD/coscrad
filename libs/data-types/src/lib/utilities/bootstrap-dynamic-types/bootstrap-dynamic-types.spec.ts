import {
    NestedDataType,
    NonEmptyString,
    NonNegativeFiniteNumber,
    Union2,
    Union2Member,
} from '../../decorators';
import { TypeDecoratorOptions } from '../../decorators/types/TypeDecoratorOptions';
import getCoscradDataSchema from '../getCoscradDataSchema';
import { bootstrapDynamicTypes } from './bootstrap-dynamic-types';

const MACHINE_UNION = 'MACHINE_UNION';

const buildOptions = (propertyName) => ({
    label: `label-for-${propertyName}`,
    description: `description-of-${propertyName}`,
});

class WhatsitProperty {
    @NonNegativeFiniteNumber(buildOptions('power'))
    power: number;

    @NonEmptyString(buildOptions('whatsit'))
    class: string;
}

@Union2Member(MACHINE_UNION, 'widget')
class Widget {
    type = 'widget';

    @NonEmptyString(buildOptions('widget'))
    name: string;
}

@Union2Member(MACHINE_UNION, 'whatsit')
class Whatsit {
    type = 'whatsit';

    name: string;

    @NestedDataType(WhatsitProperty, buildOptions('properties'))
    properties: WhatsitProperty;
}

class Undecorated {
    foo: number;

    bar: string;
}

const MachineUnion = (options: TypeDecoratorOptions) => Union2(MACHINE_UNION, 'type', options);

class FactoryRoom {
    @MachineUnion({
        label: 'machine',
        description: 'the machine that is in this factory room',
    })
    machine: Widget | Whatsit;
}

class ProductionLine {
    @MachineUnion({
        label: 'machines',
        description: `all machines that are part of this line, in order`,
    })
    machines: (Widget | Whatsit)[];
}

class HasNestedUnionProperty {
    @NestedDataType(FactoryRoom, buildOptions('factoryRoom'))
    factoryRoom: FactoryRoom;
}

type DiscriminantAndCtorName = {
    discriminant: string;
    ctorName: string;
};

const expectDataSchemaToMatchSnapshot = (ctor: unknown) => {
    expect(getCoscradDataSchema(ctor)).toMatchSnapshot();
};

describe(`bootstrapDynamicTypes`, () => {
    describe(`when several union members have been registered`, () => {
        describe.only(`when the input is valid`, () => {
            describe(`when only one class property has been decorated`, () => {
                it.only(`should append the data-schema to the metadata`, () => {
                    const initialDataSchema = getCoscradDataSchema(FactoryRoom);

                    bootstrapDynamicTypes([Whatsit, Widget, FactoryRoom, Undecorated]);

                    expectDataSchemaToMatchSnapshot(FactoryRoom);

                    const finalDataSchema = getCoscradDataSchema(FactoryRoom);

                    expect(initialDataSchema).not.toEqual(finalDataSchema);
                });
            });

            describe(`when multiple classes have a property that has been decorated`, () => {
                it(`should append the data-schema to the metadata`, () => {
                    bootstrapDynamicTypes(
                        // FactoryRoom and ProductionLine both leverage a union decorator for the same union
                        [Whatsit, Widget, Undecorated, FactoryRoom, ProductionLine]
                    );

                    expectDataSchemaToMatchSnapshot(FactoryRoom);

                    expectDataSchemaToMatchSnapshot(ProductionLine);
                });
            });

            describe(`when there is a nested property that leverages a union type`, () => {
                it(`should append the data-schema to the metadata`, () => {
                    const unionMap = bootstrapDynamicTypes(
                        // FactoryRoom and ProductionLine both leverage a union decorator for the same union
                        [
                            Whatsit,
                            Widget,
                            // Undecorated,
                            FactoryRoom,
                            // ProductionLine,
                            HasNestedUnionProperty,
                        ]
                    );

                    expectDataSchemaToMatchSnapshot(HasNestedUnionProperty);
                });
            });
        });
    });

    describe(`when no union members have been registered`, () => {
        const LONELY_UNION = 'LONELY_HEARTS_CLUB';

        class LonelyUnionUser {
            @Union2(LONELY_UNION, 'type', {
                label: 'foo',
                description: `bar can be as lonely as foo, it's the loneliest var in the village hoo`,
            })
            foo: unknown;
        }

        it(`should throw`, () => {
            const act = () =>
                bootstrapDynamicTypes([Whatsit, Widget, FactoryRoom, Undecorated, LonelyUnionUser]);

            expect(act).toThrow();
        });
    });

    describe(`when the same union has been registered twice with inconsistent discriminant paths`, () => {
        const DUPLICATED_UNION = 'DUPLICATED_UNION';

        class Woo {
            @Union2(DUPLICATED_UNION, 'type', buildOptions('foo'))
            foo: unknown;
        }

        class Boo {
            @Union2(DUPLICATED_UNION, 'inconsistentDiscriminantFieldName', buildOptions('baz'))
            baz: unknown;
        }

        @Union2Member(DUPLICATED_UNION, 'A')
        class UnionMemberA {
            inconsistentDiscriminantFieldName = 'A';
        }

        @Union2Member(DUPLICATED_UNION, 'B')
        class UnionMemberB {
            type = 'B';
        }

        it(`should throw`, () => {
            const act = () =>
                bootstrapDynamicTypes([
                    Whatsit,
                    Widget,
                    FactoryRoom,
                    Undecorated,
                    Woo,
                    Boo,
                    UnionMemberA,
                    UnionMemberB,
                ]);

            expect(act).toThrow();
        });
    });

    describe(`when two union members have registered the same discriminant value`, () => {
        // Duplicates Widget discriminant value
        @Union2Member(MACHINE_UNION, 'widget')
        class WidgetWannabe {
            type = 'widget';

            @NonEmptyString(buildOptions('name'))
            name: string;
        }

        it(`should throw`, () => {
            const act = () =>
                bootstrapDynamicTypes([Whatsit, Widget, FactoryRoom, Undecorated, WidgetWannabe]);

            expect(act).toThrow();
        });
    });
});
