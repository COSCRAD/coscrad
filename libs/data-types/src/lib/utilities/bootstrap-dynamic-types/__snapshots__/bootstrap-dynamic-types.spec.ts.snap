// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bootstrapDynamicTypes when several union members have been registered when the input is valid when multiple classes have a property that has been decorated should append the data-schema to the metadata 1`] = `
{
  "machine": {
    "complexDataType": "UNION_TYPE",
    "description": "the machine that is in this factory room",
    "discriminantPath": "type",
    "isArray": false,
    "isOptional": false,
    "label": "machine",
    "schemaDefinitions": [
      {
        "discriminantValue": "whatsit",
        "schema": {
          "properties": {
            "complexDataType": "NESTED_TYPE",
            "description": "description-of-properties",
            "isArray": false,
            "isOptional": false,
            "label": "label-for-properties",
            "schema": {
              "class": {
                "coscradDataType": "NON_EMPTY_STRING",
                "description": "description-of-whatsit",
                "isArray": false,
                "isOptional": false,
                "label": "label-for-whatsit",
              },
              "power": {
                "coscradDataType": "NON_NEGATIVE_FINITE_NUMBER",
                "description": "description-of-power",
                "isArray": false,
                "isOptional": false,
                "label": "label-for-power",
              },
            },
          },
        },
      },
      {
        "discriminantValue": "widget",
        "schema": {
          "name": {
            "coscradDataType": "NON_EMPTY_STRING",
            "description": "description-of-widget",
            "isArray": false,
            "isOptional": false,
            "label": "label-for-widget",
          },
        },
      },
    ],
    "unionName": "MACHINE_UNION",
  },
}
`;

exports[`bootstrapDynamicTypes when several union members have been registered when the input is valid when multiple classes have a property that has been decorated should append the data-schema to the metadata 2`] = `
{
  "machines": {
    "complexDataType": "UNION_TYPE",
    "description": "all machines that are part of this line, in order",
    "discriminantPath": "type",
    "isArray": false,
    "isOptional": false,
    "label": "machines",
    "schemaDefinitions": [
      {
        "discriminantValue": "whatsit",
        "schema": {
          "properties": {
            "complexDataType": "NESTED_TYPE",
            "description": "description-of-properties",
            "isArray": false,
            "isOptional": false,
            "label": "label-for-properties",
            "schema": {
              "class": {
                "coscradDataType": "NON_EMPTY_STRING",
                "description": "description-of-whatsit",
                "isArray": false,
                "isOptional": false,
                "label": "label-for-whatsit",
              },
              "power": {
                "coscradDataType": "NON_NEGATIVE_FINITE_NUMBER",
                "description": "description-of-power",
                "isArray": false,
                "isOptional": false,
                "label": "label-for-power",
              },
            },
          },
        },
      },
      {
        "discriminantValue": "widget",
        "schema": {
          "name": {
            "coscradDataType": "NON_EMPTY_STRING",
            "description": "description-of-widget",
            "isArray": false,
            "isOptional": false,
            "label": "label-for-widget",
          },
        },
      },
    ],
    "unionName": "MACHINE_UNION",
  },
}
`;

exports[`bootstrapDynamicTypes when several union members have been registered when the input is valid when only one class property has been decorated should append the data-schema to the metadata 1`] = `
{
  "machine": {
    "complexDataType": "UNION_TYPE",
    "description": "the machine that is in this factory room",
    "discriminantPath": "type",
    "isArray": false,
    "isOptional": false,
    "label": "machine",
    "schemaDefinitions": [
      {
        "discriminantValue": "whatsit",
        "schema": {
          "properties": {
            "complexDataType": "NESTED_TYPE",
            "description": "description-of-properties",
            "isArray": false,
            "isOptional": false,
            "label": "label-for-properties",
            "schema": {
              "class": {
                "coscradDataType": "NON_EMPTY_STRING",
                "description": "description-of-whatsit",
                "isArray": false,
                "isOptional": false,
                "label": "label-for-whatsit",
              },
              "power": {
                "coscradDataType": "NON_NEGATIVE_FINITE_NUMBER",
                "description": "description-of-power",
                "isArray": false,
                "isOptional": false,
                "label": "label-for-power",
              },
            },
          },
        },
      },
      {
        "discriminantValue": "widget",
        "schema": {
          "name": {
            "coscradDataType": "NON_EMPTY_STRING",
            "description": "description-of-widget",
            "isArray": false,
            "isOptional": false,
            "label": "label-for-widget",
          },
        },
      },
    ],
    "unionName": "MACHINE_UNION",
  },
}
`;

exports[`bootstrapDynamicTypes when several union members have been registered when the input is valid when there is a nested property that leverages a union type should append the data-schema to the metadata 1`] = `
{
  "factoryRoom": {
    "complexDataType": "NESTED_TYPE",
    "description": "description-of-factoryRoom",
    "isArray": false,
    "isOptional": false,
    "label": "label-for-factoryRoom",
    "schema": {
      "machine": {
        "complexDataType": "UNION_TYPE",
        "description": "the machine that is in this factory room",
        "discriminantPath": "type",
        "isArray": false,
        "isOptional": false,
        "label": "machine",
        "schemaDefinitions": [
          {
            "discriminantValue": "whatsit",
            "schema": {
              "properties": {
                "complexDataType": "NESTED_TYPE",
                "description": "description-of-properties",
                "isArray": false,
                "isOptional": false,
                "label": "label-for-properties",
                "schema": {
                  "class": {
                    "coscradDataType": "NON_EMPTY_STRING",
                    "description": "description-of-whatsit",
                    "isArray": false,
                    "isOptional": false,
                    "label": "label-for-whatsit",
                  },
                  "power": {
                    "coscradDataType": "NON_NEGATIVE_FINITE_NUMBER",
                    "description": "description-of-power",
                    "isArray": false,
                    "isOptional": false,
                    "label": "label-for-power",
                  },
                },
              },
            },
          },
          {
            "discriminantValue": "widget",
            "schema": {
              "name": {
                "coscradDataType": "NON_EMPTY_STRING",
                "description": "description-of-widget",
                "isArray": false,
                "isOptional": false,
                "label": "label-for-widget",
              },
            },
          },
        ],
        "unionName": "MACHINE_UNION",
      },
    },
  },
}
`;
